### PR TITLE
test: setup stryker for mutation testing

### PR DESCRIPTION
**Pull request by ChristianMurphy**
### Initial checklist

*   [x] I read the support docs 
*   [x] I read the contributing guide 
*   [x] I agree to follow the code of conduct 
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) 
*   [x] If applicable, I’ve added docs and tests

### Description of changes

Adds Stryker as a mutation testing tool, which tests the tests, by making breaking changes in the source code, and see if the tests still pass.

### CODE REVIEW

**Comment by ChristianMurphy** [comment]
Example output

```
All tests
  ✓ All tests (killed 151)

#0. [Survived] ArrayDeclaration
lib/index.js:47:15
-   const order = ['history', 'path', 'basename', 'stem', 'extname', 'dirname']
+   const order = []
Ran all tests for this mutant.

#1. [Survived] StringLiteral
lib/index.js:47:16
-   const order = ['history', 'path', 'basename', 'stem', 'extname', 'dirname']
+   const order = ["", 'path', 'basename', 'stem', 'extname', 'dirname']
Ran all tests for this mutant.

#3. [Survived] StringLiteral
lib/index.js:47:35
-   const order = ['history', 'path', 'basename', 'stem', 'extname', 'dirname']
+   const order = ['history', 'path', "", 'stem', 'extname', 'dirname']
Ran all tests for this mutant.

#2. [Survived] StringLiteral
lib/index.js:47:27
-   const order = ['history', 'path', 'basename', 'stem', 'extname', 'dirname']
+   const order = ['history', "", 'basename', 'stem', 'extname', 'dirname']
Ran all tests for this mutant.

#6. [Survived] StringLiteral
lib/index.js:47:66
-   const order = ['history', 'path', 'basename', 'stem', 'extname', 'dirname']
+   const order = ['history', 'path', 'basename', 'stem', 'extname', ""]
Ran all tests for this mutant.

#30. [Survived] EqualityOperator
lib/index.js:147:12
-       while (++index < order.length) {
+       while (++index <= order.length) {
Ran all tests for this mutant.

#42. [Survived] ConditionalExpression
lib/index.js:154:22
-           this[prop] = prop === 'history' ? [...options[prop]] : options[prop]
+           this[prop] = false ? [...options[prop]] : options[prop]
Ran all tests for this mutant.

#44. [Survived] StringLiteral
lib/index.js:154:31
-           this[prop] = prop === 'history' ? [...options[prop]] : options[prop]
+           this[prop] = prop === "" ? [...options[prop]] : options[prop]
Ran all tests for this mutant.

#45. [Survived] ArrayDeclaration
lib/index.js:154:43
-           this[prop] = prop === 'history' ? [...options[prop]] : options[prop]
+           this[prop] = prop === 'history' ? [] : options[prop]
Ran all tests for this mutant.

#48. [Survived] ConditionalExpression
lib/index.js:164:11
-         if (!order.includes(prop)) this[prop] = options[prop]
+         if (true) this[prop] = options[prop]
Ran all tests for this mutant.

#90. [Survived] StringLiteral
lib/index.js:241:25
-       assertPart(extname, 'extname')
+       assertPart(extname, "")
Ran all tests for this mutant.

Ran 0.99 tests per mutant on average.
----------|---------|----------|-----------|------------|----------|---------|
File      | % score | # killed | # timeout | # survived | # no cov | # error |
----------|---------|----------|-----------|------------|----------|---------|
All files |   93.25 |      151 |         1 |         11 |        0 |       0 |
 index.js |   93.25 |      151 |         1 |         11 |        0 |       0 |
----------|---------|----------|-----------|------------|----------|---------|
```

it lists how many edits it tried.
then it shows edits, which can be made, which do not fail the test suite.
then it shows some mutation coverage statistics.

**Comment by ChristianMurphy** [comment]
List has narrowed down a lot

```
All tests
  ✓ All tests (killed 157)

#6. [Survived] StringLiteral
lib/index.js:47:66
-   const order = ['history', 'path', 'basename', 'stem', 'extname', 'dirname']
+   const order = ['history', 'path', 'basename', 'stem', 'extname', ""]
Ran all tests for this mutant.

Ran 1.00 tests per mutant on average.
----------|---------|----------|-----------|------------|----------|---------|
File      | % score | # killed | # timeout | # survived | # no cov | # error |
----------|---------|----------|-----------|------------|----------|---------|
All files |   99.37 |      157 |         0 |          1 |        0 |       0 |
 index.js |   99.37 |      157 |         0 |          1 |        0 |       0 |
----------|---------|----------|-----------|------------|----------|---------|
```

I could use your insights on the last ordering test @wooorm.
I thought that
```ts
  t.test('should handle ordering of basename', (t) => {
    const file = new VFile({dirname: '~', basename: 'a.js'})

    t.deepEqual(file.history, ['a.js', path.join('~', 'a.js')])
    t.equal(file.value, undefined)
    t.equal(file.path, path.join('~', 'a.js'))
    t.equal(file.dirname, '~')
    t.equal(file.basename, 'a.js')
    t.equal(file.stem, 'a')
    t.equal(file.extname, '.js')

    t.end()
  })
```

might cover it, but it does not.
Because of

https://github.com/vfile/vfile/blob/11618f64369bb024fd1f4c2851f8ce6c46c21136/lib/index.js#L206-L209

only depending on `basename`, but `dirname` is ordered after

https://github.com/vfile/vfile/blob/11618f64369bb024fd1f4c2851f8ce6c46c21136/lib/index.js#L47

so it could never do anything? Am I missing an edge case? Could `dirname` be removed from the list?

File: test.js [hunk 1]
```diff
@@ -153,6 +153,47 @@ test('new VFile(options?)', (t) => {
...
+    const file = new VFile({
+      history: externalHistory
+    })
+
+    t.deepEqual(file.history, ['a.js', path.join('~', 'a.js')])
+    externalHistory[0] = ''
+    t.deepEqual(file.history, ['a.js', path.join('~', 'a.js')])
+
+    t.end()
+  })
```
Comment (ChristianMurphy) [inline test.js:195]: this squashes

```
#42. [Survived] ConditionalExpression
lib/index.js:154:22
-           this[prop] = prop === 'history' ? [...options[prop]] : options[prop]
+           this[prop] = false ? [...options[prop]] : options[prop]
Ran all tests for this mutant.

#44. [Survived] StringLiteral
lib/index.js:154:31
-           this[prop] = prop === 'history' ? [...options[prop]] : options[prop]
+           this[prop] = prop === "" ? [...options[prop]] : options[prop]
Ran all tests for this mutant.

#45. [Survived] ArrayDeclaration
lib/index.js:154:43
-           this[prop] = prop === 'history' ? [...options[prop]] : options[prop]
+           this[prop] = prop === 'history' ? [] : options[prop]
```

and

```
#1. [Survived] StringLiteral
lib/index.js:47:16
-   const order = ['history', 'path', 'basename', 'stem', 'extname', 'dirname']
+   const order = ["", 'path', 'basename', 'stem', 'extname', 'dirname']
Ran all tests for this mutant.
```

File: test.js [hunk 2]
```diff
@@ -153,6 +153,47 @@ test('new VFile(options?)', (t) => {
...
+    t.deepEqual(file.history, [path.join('~', 'b.js'), path.join('~', 'a.js')])
+    t.equal(file.value, undefined)
+    t.equal(file.path, path.join('~', 'a.js'))
+    t.equal(file.dirname, '~')
+    t.equal(file.basename, 'a.js')
+    t.equal(file.stem, 'a')
+    t.equal(file.extname, '.js')
+
+    t.end()
+  })
```
Comment (ChristianMurphy) [inline test.js:168]: squashes

```
#2. [Survived] StringLiteral
lib/index.js:47:27
-   const order = ['history', 'path', 'basename', 'stem', 'extname', 'dirname']
+   const order = ['history', "", 'basename', 'stem', 'extname', 'dirname']
Ran all tests for this mutant.
```

File: test.js [hunk 3]
```diff
@@ -153,6 +153,47 @@ test('new VFile(options?)', (t) => {
...
+    t.deepEqual(file.history, ['a.js', path.join('~', 'a.js')])
+    t.equal(file.value, undefined)
+    t.equal(file.path, path.join('~', 'a.js'))
+    t.equal(file.dirname, '~')
+    t.equal(file.basename, 'a.js')
+    t.equal(file.stem, 'a')
+    t.equal(file.extname, '.js')
+
+    t.end()
+  })
```
Comment (ChristianMurphy) [inline test.js:182]: squashes

```
#3. [Survived] StringLiteral
lib/index.js:47:35
-   const order = ['history', 'path', 'basename', 'stem', 'extname', 'dirname']
+   const order = ['history', 'path', "", 'stem', 'extname', 'dirname']
Ran all tests for this mutant.
```

File: test.js [hunk 4]
```diff
@@ -392,7 +433,15 @@ test('new VFile(options?)', (t) => {
...
-      'should throw with mutiple `.`s'
+      'should throw with multiple `.`s'
+    )
+
+    t.throws(
+      () => {
+        file.extname = path.sep + ".md"
+      },
+      /Error: `extname` cannot be a path: did not expect `.`/,
+      'should throw with path separator included'
```
Comment (ChristianMurphy) [inline test.js:444]: squashes

```
#90. [Survived] StringLiteral
lib/index.js:241:25
-       assertPart(extname, 'extname')
+       assertPart(extname, "")
Ran all tests for this mutant.
```

**Comment by ChristianMurphy** [comment]
List has narrowed down a lot

```
All tests
  ✓ All tests (killed 157)

#6. [Survived] StringLiteral
lib/index.js:47:66
-   const order = ['history', 'path', 'basename', 'stem', 'extname', 'dirname']
+   const order = ['history', 'path', 'basename', 'stem', 'extname', ""]
Ran all tests for this mutant.

Ran 1.00 tests per mutant on average.
----------|---------|----------|-----------|------------|----------|---------|
File      | % score | # killed | # timeout | # survived | # no cov | # error |
----------|---------|----------|-----------|------------|----------|---------|
All files |   99.37 |      157 |         0 |          1 |        0 |       0 |
 index.js |   99.37 |      157 |         0 |          1 |        0 |       0 |
----------|---------|----------|-----------|------------|----------|---------|
```

I could use your insights on the last ordering test @wooorm.
I thought that
```ts
  t.test('should handle ordering of basename', (t) => {
    const file = new VFile({dirname: '~', basename: 'a.js'})

    t.deepEqual(file.history, ['a.js', path.join('~', 'a.js')])
    t.equal(file.value, undefined)
    t.equal(file.path, path.join('~', 'a.js'))
    t.equal(file.dirname, '~')
    t.equal(file.basename, 'a.js')
    t.equal(file.stem, 'a')
    t.equal(file.extname, '.js')

    t.end()
  })
```

might cover it, but it does not.
Because of

https://github.com/vfile/vfile/blob/11618f64369bb024fd1f4c2851f8ce6c46c21136/lib/index.js#L206-L209

only depending on `basename`, but `dirname` is ordered after

https://github.com/vfile/vfile/blob/11618f64369bb024fd1f4c2851f8ce6c46c21136/lib/index.js#L47

so it could never do anything? Am I missing an edge case? Could `dirname` be removed from the list?

File: lib/index.js [hunk 5]
```diff
@@ -142,11 +142,7 @@ export class VFile {
     /* eslint-enable no-unused-expressions */
 
     // Set path related properties in the correct order.
-    let index = -1
-
-    while (++index < order.length) {
-      const prop = order[index]
-
+    for (const prop of order) {
```
Comment (ChristianMurphy) [inline lib/index.js:145]: squashes

```
#30. [Survived] EqualityOperator
lib/index.js:147:12
-       while (++index < order.length) {
+       while (++index <= order.length) {
Ran all tests for this mutant.
```

**Comment by ChristianMurphy** [comment]
trying some more variations to maybe test `dirname` order, still nothing seems to break if it is removed from order

<details>
<summary>tests tried</summary>

```ts
  t.test('should handle ordering of dirname', (t) => {
    const file = new VFile({dirname: '~', stem: 'example', extname: '.md'})

    t.deepEqual(file.history, [
      'example',
      'example.md',
      path.join('~', 'example.md')
    ])
    t.equal(file.value, undefined)
    t.equal(file.path, path.join('~', 'example.md'))
    t.equal(file.dirname, '~')
    t.equal(file.basename, 'example.md')
    t.equal(file.stem, 'example')
    t.equal(file.extname, '.md')

    t.end()
  })
```

```ts
  t.test('should handle ordering of dirname', (t) => {
    const file = new VFile({path: '~/example.md', dirname: 'etc'})

    t.deepEqual(file.history, [
      path.join('~', 'example.md'),
      path.join('etc', 'example.md')
    ])
    t.equal(file.value, undefined)
    t.equal(file.path, path.join('etc', 'example.md'))
    t.equal(file.dirname, 'etc')
    t.equal(file.basename, 'example.md')
    t.equal(file.stem, 'example')
    t.equal(file.extname, '.md')

    t.end()
  })
```

```ts
  t.test('should handle ordering of dirname', (t) => {
    const file = new VFile({dirname: 'etc', path: '~/example.md'})

    t.deepEqual(file.history, [
      path.join('~', 'example.md'),
      path.join('etc', 'example.md')
    ])
    t.equal(file.value, undefined)
    t.equal(file.path, path.join('etc', 'example.md'))
    t.equal(file.dirname, 'etc')
    t.equal(file.basename, 'example.md')
    t.equal(file.stem, 'example')
    t.equal(file.extname, '.md')

    t.end()
  })
```

</details>

File: stryker.conf.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{
+  "$schema": "./node_modules/@stryker-mutator/core/schema/stryker-schema.json",
+  "packageManager": "npm",
+  "reporters": ["html", "clear-text", "progress"],
+  "mutate": ["lib/index.js"],
```
Comment (ChristianMurphy) [inline stryker.conf.json:5]: only mutate the index file, the browser files are copied from node/browserify

File: stryker.conf.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+  "testRunner": "command",
+  "coverageAnalysis": "off",
+  "commandRunner": {"command": "npm run test-api"},
```
Comment (ChristianMurphy) [inline stryker.conf.json:8]: some test frameworks have coverage aware runners https://www.npmjs.com/search?q=%40stryker-mutator%20runner
`tape` currently isn't supported, so we fallback to the `command` runner which runs all tests every time

File: stryker.conf.json [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+  "timeoutFactor": 3,
+  "thresholds": {"break": 100}
```
Comment (ChristianMurphy) [inline stryker.conf.json:10]: fails the build if any mutations survive without failing the build

File: lib/index.js (hunk 5 above)
Comment (wooorm) [reply lib/index.js:145]: I don’t get what this solves 🤷‍♂️ Why would a broken loop suddenly be created.

**Review by wooorm — COMMENTED** [review]
I’m not sure. This feels too much to me. E.g., the loop is really fine, the `order` has been the same for 6 years already without a problem. I feel like copying `history` instead of referencing it is really fine without an explicit test.

**Comment by wooorm** [comment]
> so it could never do anything? Am I missing an edge case? Could dirname be removed from the list?

No. The order is not just about `basename` and `dirname` (the two adjacent last values). The order is also about the first (say, `path`) and the last (`dirname`). Replacing `basename` in your test with some other earlier values should do the trick

File: lib/index.js (hunk 5 above)
Comment (ChristianMurphy) [reply lib/index.js:145]: Mutation testing is testing the tests.
Here it found that the condition in the loop could be modified without breaking a test, a signal that there is potential code.
We could put an ignore on the line https://github.com/stryker-mutator/stryker-js/blob/bfb33030b10f8c95fd5a7bb1a9307881b50629f4/docs/disable-mutants.md#using-a--stryker-disable-comment
But here it felt cleaner to use a `for of` loop instead, since the intent is to loop over the elements in the list.

**Comment by ChristianMurphy** [comment]
> I’m not sure. This feels too much to me. E.g., the loop is really fine, the `order` has been the same for 6 years already without a problem. I feel like copying `history` instead of referencing it is really fine without an explicit test.

I'm not saying it's broken, it works swell and I'd like to keep it that way.

What the mutator is zeroing in on, are behaviors which you felt were important enough to code into the logic, but are not covered by the test suite.
It feels strange to push for 100% test coverage, but to be okay with the tests not covering key parts of the behavior.

**Comment by ChristianMurphy** [comment]
It's also worth calling out that there are a few approaches that can be taken a few fronts.

In terms of when to run it:
1. always and automatically (currently set)
2. occasionally and manually
3. never

In terms of what to check
1. ensure :100: coverage (currently set)
2. ensure a lower coverage number
3. don't set a minimum coverage level at all

**Comment by remcohaszing** [comment]
I’m not really sure if I see the added value of mutation testing.

@wooorm wrote:

> I’m not sure. This feels too much to me. E.g., the loop is really fine, the order has been the same for 6 years already without a problem. I feel like copying history instead of referencing it is really fine without an explicit test.

:point_up_2: This exactly describes my thoughts on this.

File: lib/index.js (hunk 5 above)
Comment (wooorm) [reply lib/index.js:145]: To expand, on this change: it’s the essence of why I’m 🤷‍♂️ / :-1: on this PR / mutation tests.

- What Stryker reports is confusing: for me personally, but presumably also for anyone that wants to contribute
- This “solution” is *different*, it doesn’t solve anything except for Stryker

I’m fine with all the added test cases, somewhat verbose, but why not.
But this change is IMO a regression that’s `stryker ignore`ing problems in Stryker itself.

For Stryker more generally: the added tooling to me feels like the direction the MDX repo was heading into a while ago. Too much tooling, too slow, constantly broken and unusable, and hard to actually contribute code to.

I feel like TypeScript and the xo direction already result in much of that. A lot of busy work / churn without benefits.

---

> This is basically a masked `for` loop.

If `xo` starts disallowed `while` loops I would personally yeet it into the trash on every project I can.
I find `while` loops are the most powerful and clear loops that exist. The essence of a loop. `for of` loops have their use but IMO don’t need to be used solely.

**Comment by ChristianMurphy** [comment]
> Isn’t `extname` the same btw?

it changes the behavior of history, for example

```js
  t.test('should accept a object (2)', (t) => {
    const file = new VFile({stem: 'example', extname: '.md', dirname: '~'})

    t.deepEqual(file.history, [
      'example',
      'example.md',
      path.join('~', 'example.md')
    ])
    t.equal(file.value, undefined)
    t.equal(file.path, path.join('~', 'example.md'))
    t.equal(file.dirname, '~')
    t.equal(file.basename, 'example.md')
    t.equal(file.stem, 'example')
    t.equal(file.extname, '.md')

    t.end()
  })
```

with `extname` removed from the `order` array would produce:

```markdown
# should accept a object (2)
not ok 31 should be deeply equivalent
  ---
    operator: deepEqual
    expected: [ 'example', 'example.md', '~/example.md' ]
    actual:   [ 'example', '~/example', '~/example.md' ]
```

File: lib/index.js (hunk 5 above)
Comment (ChristianMurphy) [reply lib/index.js:145]: > What Stryker reports is confusing: for me personally, but presumably also for anyone that wants to contribute

Is there anything that could be added to make it clearer?

In other conversations, we've touched on the topic that mutation testing, can be thought of roughly as a type of test coverage.
If a contributor came through and said "line coverage is confusing, so I didn't add it" you'd close the PR without further discussion.

> This “solution” is different, it doesn’t solve anything except for Stryker

It improves readability, Stryker drew my attention to this line. I would have suggested this change regardless of the coverage difference.

> But this change is IMO a regression

I disagree, readable code is not a regression.

> that’s stryker ignore ing problems in Stryker itself.

Not really, the mutation test is doing what it's supposed to, seeing if editing code breaks things.

> Too much tooling, too slow, constantly broken and unusable,

breaking this down, and focusing this back on vfile, rather than pent up frustration with mdx

* too much tooling: this is highly subjective and I respectfully disagree
* too slow: it takes a minute and thirty seconds total for all tests including mutation tests
* constantly broken: this is a type of test coverage, if the code isn't covered by tests, the build should be broken :thinking: 
* unusable: again highly subjective, I find both vfile and mdx are quite usable

> and hard to actually contribute code to.

in MDX that was more governance related than tooling related

Comment (remcohaszing) [reply lib/index.js:145]: Are Stryker results predictable? As in, does it always make the same mutations? 

In my experience it’s frustrating if CI fails due to factors other than the changes made.

Comment (ChristianMurphy) [reply lib/index.js:145]: > Are Stryker results predictable? As in, does it always make the same mutations?

It does have predictable results, it has a fixed number of mutators https://stryker-mutator.io/docs/mutation-testing-elements/supported-mutators
and it runs all enabled mutators through all replacement combinations, on every enabled file.

**Comment by ChristianMurphy** [comment]
> I’m not really sure if I see the added value of mutation testing.

Another way of conceptualizing it, is a different type of code coverage measure.

* line coverage
   - sees: if a line was run
   - misses: if a static or dynamic condition was run, if there is an assertion which breaks if the code does
* branch coverage
   - sees: if a static branch was run
   - misses: if a dynamic condition was run, and if there is an assertion which breaks if the code does
* mutation coverage
    - sees: if changing the code breaks the tests, this includes static and dynamic conditions
    - misses: if a line was run or not, and code which cannot be mutated https://stryker-mutator.io/docs/mutation-testing-elements/supported-mutators

these types of coverage compliment each other.
with mutation testing able to find conditions in the code which have gone untested for years.

> > I’m not sure. This feels too much to me. E.g., the loop is really fine, the order has been the same for 6 years already without a problem. I feel like copying history instead of referencing it is really fine without an explicit test.
>
> :point_up_2: This exactly describes my thoughts on 

:man_shrugging: the way I read this is: "it just so happened that nobody has edited this untested code in 6 years, so let's leave it untested"
Which is an okay opinion to have.

It does seem to run counter to the goal of having the project be easy for developers to contribute to.
I'm not certain a new contributor would be able to piece together what is intentionally supported.
I have contributed to this project before, and was not aware that these cases need to be supported.

**Comment by ChristianMurphy** [comment]
building on https://github.com/vfile/vfile/pull/70#issuecomment-1004396565

another potential option would be to leave the tests in this PR.
But move the process that runs the mutator out.
It could be a separate job either in this repo, or run from another repo.
And could be set to run once every few months, or be run manually before a major release.

File: package.json [hunk 9]
```diff
@@ -74,7 +75,8 @@
     "format": "remark . -qfo && prettier . -w --loglevel warn && xo --fix",
     "test-api": "node test.js",
     "test-coverage": "c8 --check-coverage --branches 100 --functions 100 --lines 100 --statements 100 --reporter lcov node test.js",
-    "test": "npm run build && npm run format && npm run test-coverage"
+    "test-mutation": "npx stryker run",
```
Comment (remcohaszing) [inline package.json:78]: `node_modules/.bin` is already in `PATH` in npm scripts, so there’s no need to use `npx` here.

```suggestion
    "test-mutation": "stryker run",
```

**Review by remcohaszing — CHANGES_REQUESTED** [review]

File: lib/index.js [hunk 10]
```diff
@@ -44,7 +44,8 @@ import {urlToPath, isUrl} from './minurl.js'
 // Order of setting (least specific to most), we need this because otherwise
 // `{stem: 'a', path: '~/b.js'}` would throw, as a path is needed before a
 // stem can be set.
-const order = ['history', 'path', 'basename', 'stem', 'extname', 'dirname']
+// Note: extname and dirname automatically run last
```
Comment (wooorm) [inline lib/index.js:47]: This comment is not formulated correctly. Anything in `options` not in this list “automatically” runs last, although it is in fact rather manually, there’s code, after the loop, to do that.

It is important that `extname` and `dirname` run after `history`, `path`, `basename`, and `stem`. Explicitly, as they have been amongst their peers, or implicitly by removing them.

I think it’s rather weird that not all special path-related props are listed now. This can be changed by changing the variable name and the whole comment.

File: lib/index.js (hunk 5 above)
Comment (wooorm) [reply lib/index.js:146]: > Is there anything that could be added to make it clearer?

istanbul/c8/etc reports are fine.

> In other conversations, we've touched on the topic that mutation testing, can be thought of roughly as a type of test coverage.

I frame this differently. There are different types of tests: unit tests, integration, mutation, fuzzing, etc. They all have coverage. The unit tests in all of unified have 100% test coverage. Adding mutation tests don’t increase that number. It’s a different metric. It adds more tests.

> If a contributor came through and said "line coverage is confusing, so I didn't add it" you'd close the PR without further discussion.

Yes. I don’t want to do that for mutation tests.
Adding more tools doesn’t help this point. Not adding more tools helps this. Having some tools already doesn’t change this.

> It improves readability [... and the rest]

I disagree on `for .. of` loops increasing readability over `while`.

> rather than pent up frustration with mdx

It is an example. It is a different example from what you are responding to though. In 2020, the MDX repo was unusable.

> too much tooling: this is highly subjective and I respectfully disagree

I was explaining problems with MDX 2 years ago. But to expand on Stryker:
- Mutation testing is not common or standard practise compared to say unit tests or coverage.
- OG `stryker` and styker/core are used in very few repos after 5 years.
- It doesn’t really help in this PR. The code is fine. A while loop is changed.

> too slow: it takes a minute and thirty seconds total for all tests including mutation tests

This was about MDX. For Stryker: stryker/core has a 30mb install size. It does run the tests 100(s) of times so that costs some time.
This was mostly about adding several, arbitrary tools though.

> constantly broken: this is a type of test coverage, if the code isn't covered by tests, the build should be broken 🤔

This was in reference to MDX. For Stryker: i’m worried that it’ll start emitting an unreadable message for every (other) line that’s added.
This was mostly about adding several, arbitrary tools though, because tools break.

**Comment by ChristianMurphy** [comment]
> istanbul/c8/etc reports are fine.
> ...
> For Stryker: i’m worried that it’ll start emitting an unreadable message for every (other) line that’s added.

It also generates an HTML report if that's what you mean.
It's mostly useful locally, but we could configure CI to make it a temporary artifact so it can be easily accessed from there as well.

> The unit tests in all of unified have 100% test coverage

Line and branch coverage specifically, yes

> It’s a different metric

Right

> Adding more tools doesn’t help this point. Not adding more tools helps this. Having some tools already doesn’t change this.

I respectfully disagree.
That puts all the onus on code review, rather than getting support from automation.
If anything is missed by code review, the process of recovery from that is far more unpleasant than a red checkmark on CI.

> It is a different example from what you are responding to though. In 2020, the MDX repo was unusable.

No, we're talking about the same repo and the same time span.
It was annoying, but far from unusable.

> Mutation testing is not common or standard practice compared to say unit tests or coverage

Right, testing is a spectrum, mutation testing is further along that spectrum.
But the way you frame a strange way of looking at it.
Code quality practices on a project with 6,000,000 downloads a week are likely different from those of a project with 60 downloads a week, much less one not published at all.
It will, out of necessity invent more in code quality practices.

> It doesn’t really help in this PR. The code is fine. A while loop is changed.

I agree that the code is fine.
And that is entirely why this PR exists and helps.
There is intended behavior which is untested, anyone coming along interested in improving or fixing a regression, could introduce a regression in one of those areas without knowing.
This helps to address some of those gaps.

> This was mostly about adding several, arbitrary tools though.

It is a single tool, and not any more arbitrary than C8 or Istanbul.

**Comment by wooorm** [comment]
> It also generates an HTML report if that's what you mean.

I think the report is unclear in signalling what has to happen to silence it. This is different from linting, tests, normal coverage, which are more common / clear / documented.
I do not understand how to, from “`<=` survived”, derive that apparently while loops should not be used. It requires learning a jargon. Reading an error message and guessing what a solution might be.

> That puts all the onus on code review, rather than getting support from automation.

Indeed. I think that’s fine. I think code review is an important human process which cannot and does not have to be fully automated. I think that more tools makes it harder to contribute.

> It was annoying, but far from unusable.

I speak from how I experience things and what I heard from others. The team lead and I found the repo unusable. See also: https://github.com/mdx-js/mdx/issues/1015.

> And that is entirely why this PR exists and helps.

I don’t think these tests help. I don’t think this tool helps.

Good tests, in my opinion, assert that with a certain input, a certain actual output of a program is yielded. They allow a refactor or (to some reasonable extent) rewrite of internals.
Bad test assert the internals. They need to be changed when the internals change.
Good tests help and save time, bad tests get in the way and cost time.
(It’s not a clear distinction in practise of course).
(I’m reminded of https://kentcdodds.com/blog/write-tests, mostly the tweet, but the article too).

The added tests are *not bad* on their own, sure, they make some sense! But what you call gaps I see as breathing room for the internals.
The changes to src *are bad* in my opinion. The `while` loop and `dirname` in `order` are implementation details that tools should not crash on and do not need to be locked down with tests.
This tool pushes for tests that are coupled with the internals of the code.
It pushes for tests that have to change when the implementation does, and an implementation that has to change when the tests do.
I don’t think we need that. I think we get far enough with normal tests and test coverage.

**Comment by ChristianMurphy** [comment]
> I think the report is unclear in signalling what has to happen to silence it. This is different from linting, tests, normal coverage, which are more common / clear / documented.

I respectfully disagree, it's interpretation is largely the same as branch coverage.

> I do not understand how to, from “<= survived”, derive that apparently while loops should not be used. It requires learning a jargon. Reading an error message and guessing what a solution might be.

Again, it's interpretation is roughly the same as branch coverage.
For example, the while loop
```js
while (++index < order.length) {
```

or

```js
while (++index <= order.length) {
```

both pass unit tests. mainly because of the

```js
if (prop in options && options[prop] !== undefined) {
```
 a few lines down.
There were a few options:
1. ignore this
2. change the condition further down
3. use a for of loop

again, the choice of option 3 what mostly stylistic and for readability.

> Indeed. I think that’s fine. I think code review is an important human process which cannot and does not have to be fully automated. I think that more tools makes it harder to contribute.

I agree that it cannot be fully automated, but I think some important aspects of ensure existing and new tests cover the code can be partially automated.
In terms of ease of contribution, I would reiterate

> If anything is missed by code review, the process of recovery from that is far more unpleasant than a red checkmark on CI.

Learning how to read coverage is easier for a potential contributor than dealing with fallout from anything missed or incorrectly semver'ed.

> I speak from how I experience things and what I heard from others

And I speak from how I experience things, having maintained the repo, while you and the lead maintainer stepped away for months.

> Good tests, in my opinion, assert that with a certain input, a certain actual output of a program is yielded. They allow a refactor or (to some reasonable extent) rewrite of internals.
> Bad test assert the internals. They need to be changed when the internals change.
> Good tests help and save time, bad tests get in the way and cost time.

There are actual results of the program which are not tested.
Which this adds tests for, note that it doesn't really care what the internals are.
It cares that visible outputs are tested, and that internals have some purpose.

> The added tests are not bad on their own, sure, they make some sense! But what you call gaps I see as breathing room for the internals.

The fact that you have concerns on the potential result of changing so of these "breathing room for the internals" tells me they aren't actually internal.

> The changes to src are bad in my opinion [...] The while loop

For the umpteenth time, it is a readability change, which I would have suggested regardless of stryker.

> `dirname` in order are implementation details that tools should not crash on and do not need to be locked down with tests.

:man_shrugging: ESLint recommending removing dead code is good, but stryker doing the same is bad? :thinking: 
I respectfully disagree. Removing dead code benefits the codebase.

**Comment by wooorm** [comment]
It seems clear already, but I’ll say it explicitly:

- I’m not in favor of removing `while` loops. `for..of` is not more or less readable than `while`. I find it less readable.
- I’m not in favor of adding more, stricter tests as demonstrated here and in unifiedjs/unified#183 compared to the 100% we already have. unified is already very well tested compared to the rest of npm. I don‘t think unified benefits from more tests. I believe more tests will make things harder.
- I’m not in favor of adding tools that do not bring much value. Stryker is an uncommon tool. Mutation testing is uncommon. Contributors are unfamiliar with both. It has a big dependency graph. It does not seem to find any/many bugs here and in unified.
